File: plugin/src/main/java/com/jtprince/bingo/plugin/automarking/AutomatedGoalList.java
package com.jtprince.bingo.plugin.automarking;

import com.jtprince.bingo.plugin.automarking.itemtrigger.ItemTriggerYaml;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

/**
 * Standalone executable that lists all goals that are automated by the plugin.
 */
public class AutomatedGoalList {
    public static void main(String[] args) throws IOException {
        Set<String> goals = getAutomatedGoals();
        if (args.length > 0) {
            BufferedWriter out = new BufferedWriter(new FileWriter(args[0], false));
            for (String goal : goals) {
                out.write(goal + '\n');
            }
            out.close();
        } else {
            goals.forEach(System.out::println);
        }
    }

    static Set<String> getAutomatedGoals() {
        HashSet<String> ret = new HashSet<>();
        ret.addAll(EventTrigger.allAutomatedGoals());
        ret.addAll(OccasionalTrigger.allAutomatedGoals());
        ret.addAll(ItemTriggerYaml.defaultYaml().allAutomatedGoals());
        return ret;
    }
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/automark/AutomatedGoalList.kt
@file:JvmName("AutomatedGoalList")

package com.jtprince.bingo.kplugin.automark

import java.io.File
import java.io.PrintWriter

/**
 * Standalone executable that lists all goals that are automated by the plugin.
 *
 * If executed with an argument, goals will be written to the file specified by that argument.
 * Otherwise, goals will be printed to stdout.
 */
fun main(args: Array<String>) {
    val allGoals: Set<String> = AutoMarkTrigger.allAutomatedGoals
    val file = if (args.isNotEmpty()) File(args[0]).printWriter() else PrintWriter(System.out)

    file.use {
        for (goal in allGoals) {
            it.println(goal)
        }
    }
}
